packages/@tailwindcss-postcss/src/index.ts

```diff
@@ -36,20 +36,20 @@ export type PluginOptions = {
   optimize?: boolean | { minify?: boolean }
 }
 
+let cache = new DefaultMap(() => {
+  return {
+    mtimes: new Map<string, number>(),
+    compiler: null as null | Awaited<ReturnType<typeof compile>>,
+    css: '',
+    optimizedCss: '',
+    fullRebuildPaths: [] as string[],
+  }
+})
```

Comment:
The only thing I can think of here is that previously this was scoped for the `opts` argument to the `tailwindcss` function and now it's not, so now the cache is shared for the same key even if the options are different.

Interestingly the [PostCSS docs explicitly say](https://postcss.org/docs/writing-a-postcss-plugin) to set up caches in the place we were setting it up before:

<img width="1071" alt="image" src="https://github.com/user-attachments/assets/3a8de8c7-bcd8-46bd-9966-d32338bb0903">

Do we know with absolute certainty it was getting re-created from scratch on every save? 🤔

Reply:
If I apply this diff to our plugin:
```diff
diff --git a/packages/@tailwindcss-postcss/src/index.ts b/packages/@tailwindcss-postcss/src/index.ts
index 7d18ddc36..43b824d2f 100644
--- a/packages/@tailwindcss-postcss/src/index.ts
+++ b/packages/@tailwindcss-postcss/src/index.ts
@@ -50,6 +50,8 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
   let base = opts.base ?? process.cwd()
   let optimize = opts.optimize ?? process.env.NODE_ENV === 'production'
 
+  console.count('Setup caches here')
+
   return {
     postcssPlugin: '@tailwindcss/postcss',
     plugins: [
@@ -65,6 +67,7 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
           let inputFile = result.opts.from ?? ''
           let context = cache.get(inputFile)
           let inputBasePath = path.dirname(path.resolve(inputFile))
+          console.count('Run @tailwindcss/postcss')
 
           async function createCompiler() {
             env.DEBUG && console.time('[@tailwindcss/postcss] Setup compiler')
```

If I then use it, the output looks like this:
<img width="478" alt="image" src="https://github.com/user-attachments/assets/064db8a0-b111-4fc2-876f-0f53c3cdd006">

You have the initial build, then 2 builds because of watch mode and because I made 2 changes.

Some notes / observations:
1. Not only is the `Setup caches here` log called every time we save, it's called twice for an unknown reason.
2. This is using the `postcss-cli` package

Next steps are to verify if this behaviour also happens in other environments where the PostCSS plugin is being used.

Reply:
Can confirm that this behaviour also exists in a `vite` project with the PostCSS plugin:
<img width="822" alt="image" src="https://github.com/user-attachments/assets/e24457d3-5b4d-4ee9-8898-dfac99ca27c5">

Reply:
Some updates for completeness:

If you structure your PostCSS config like this:
```js
module.exports = {
  plugins: {
    '@tailwindcss/postcss': {},
  },
}
```
or this:
```js
module.exports = {
  plugins: [
    require('@tailwindcss/postcss')
  ],
}
```
Then the above behavior can be observer. If you _call_ your function (even without options) then you see the correct behavior:
```js
module.exports = {
  plugins: [
    require('@tailwindcss/postcss')()
  ],
}
```